### PR TITLE
Upgrade to numpy 2.0

### PR DESCRIPTION
**Pull request by suzhoum**
*Issue #, if available:*

*Description of changes:*
This PR adds support for `numpy` 2.x.x. 
1. Fixed a few compatibility issues during upgrade.
2. Capped `numpy` to use 1.x.x when `catboost` is installed (`pip install "autogluon.tabular[all]"` or `pip install autogluon.timeseries`) due to existing issue https://github.com/catboost/catboost/issues/2671

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py [hunk 1]
```diff
@@ -655,7 +655,7 @@ def transform_document(
     def transform_numerical(
         self,
         df: pd.DataFrame,
-    ) -> Tuple[Dict[str, NDArray], None]:
+    ) -> Tuple[Dict[str, NDArray[np.float64]], None]:
```
Comment (tonyhoo) [inline multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py:658]: Do we need to specify precision here? I'm curious if there will be cases where float32 is used.

File: multimodal/src/autogluon/multimodal/data/process_document.py [hunk 2]
```diff
@@ -250,8 +250,8 @@ def get_ocr_features(self, doc_path, doc_image):
 
     def process_one_sample(
         self,
-        document_features: Dict[str, Union[NDArray, list]],
-        feature_modalities: Dict[str, Union[NDArray, list]],
+        document_features: Dict[str, Union[NDArray[np.float64], list]],
```
Comment (tonyhoo) [inline multimodal/src/autogluon/multimodal/data/process_document.py:253]: Sama here

File: tabular/setup.py [hunk 3]
```diff
@@ -39,8 +39,8 @@
         # CatBoost wheel build is not working correctly on darwin for CatBoost 1.2, so use old version in this case.
         # https://github.com/autogluon/autogluon/pull/3190#issuecomment-1540599280
         # Catboost 1.2 doesn't have wheel for python 3.11
-        "catboost>=1.1,<1.2 ; sys_platform == 'darwin' and python_version < '3.11'",
-        "catboost>=1.1,<1.3; sys_platform != 'darwin'",
+        "numpy>=1.25,<2.0.0",  # TODO support numpy>=2.0.0 once issue resolved https://github.com/catboost/catboost/issues/2671
```
Comment (tonyhoo) [inline tabular/setup.py:42]: So we are not upgrading to numpy 2.0 per se?

File: timeseries/setup.py [hunk 4]
```diff
@@ -30,7 +30,7 @@
     "pytorch_lightning",  # version range defined in `core/_setup_utils.py`
     "transformers[sentencepiece]",  # version range defined in `core/_setup_utils.py`
     "accelerate",  # version range defined in `core/_setup_utils.py`
-    "gluonts==0.15.1",
```
Comment (tonyhoo) [inline timeseries/setup.py:33]: Shall we pin the version or introduce it as a lower, upper bound range?

Comment (suzhoum) [reply timeseries/setup.py:33]: I'm planning to update to lower bound once `gluonts` is release (hopefully before AG 1.2.

File: multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py (hunk 1 above)
Comment (suzhoum) [reply multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py:658]: I think this can be removed.

File: timeseries/setup.py (hunk 4 above)
Comment (suzhoum) [reply timeseries/setup.py:33]: Updated as `gluonts` just released 0.16.0

File: tabular/setup.py (hunk 3 above)
Comment (prateekdesai04) [reply tabular/setup.py:42]: are we then just merging this PR and waiting for Catboost update?
Is there significant difference in any scores from 3.9, 3.10, 3.11 compared with 3.12 in tabular?

**Review by prateekdesai04 — APPROVED** [review]

File: features/src/autogluon/features/binning.py [hunk 5]
```diff
@@ -57,7 +57,7 @@ def generate_bins(X_features: DataFrame, features_to_bin: list, ideal_bins: int
                 is_satisfied = True
                 # print('max_iterations met, stopping prior to satisfaction!', column, len(interval_index))
 
-        bins_final = interval_index.right.values
+        bins_final = np.array(interval_index.right.values, dtype=np.float64)
```
Comment (Innixma) [inline features/src/autogluon/features/binning.py:60]: Avoid using float64, it uses too much memory. Keep as whatever it was before (int?)

File: features/src/autogluon/features/generators/text_special.py [hunk 6]
```diff
@@ -138,23 +138,23 @@ def _generate_text_special(self, X: Series, feature: str, symbols: list, X_dict:
         X_dict[fn["char_count"]] = char_count.to_numpy(dtype=np.uint32)
         X_dict[fn["word_count"]] = X_str.str.split().str.len().to_numpy(dtype=np.uint32)
         X_dict[fn["capital_ratio"]] = (
-            X_no_ws.str.count("[A-Z]").divide(X_no_ws_text_len, fill_value=0.0).fillna(0.0).to_numpy(dtype=np.float32)
+            X_no_ws.str.count("[A-Z]").divide(X_no_ws_text_len, fill_value=0.0).fillna(0.0).to_numpy(dtype=np.float64)
```
Comment (Innixma) [inline features/src/autogluon/features/generators/text_special.py:141]: Avoid float64, keep float32, adjust tests accordingly. We don't want to double memory usage. Ditto for rest of cases

File: common/setup.py [hunk 7]
```diff
@@ -28,14 +28,16 @@
         "tqdm",  # version range defined in `core/_setup_utils.py`
         # s3fs is removed due to doubling install time due to version range resolution
         # "s3fs",  # version range defined in `core/_setup_utils.py`
-        "setuptools",
+        "setuptools<60.0; python_version < '3.12'",
+        "setuptools>=60.0; python_version >= '3.12'",
```
Comment (Innixma) [inline common/setup.py:32]: Do we need setuptools?

Comment (suzhoum) [reply common/setup.py:32]: Upon testing, I think it can be removed.

File: features/src/autogluon/features/binning.py (hunk 5 above)
Comment (Innixma) [reply features/src/autogluon/features/binning.py:60]: This is still float64. What was it before?

**Review by Innixma — CHANGES_REQUESTED** [review]
Refer to comments in the `binning.py` file. This seems not to be addressed.

File: tabular/setup.py [hunk 8]
```diff
@@ -36,11 +36,8 @@
         "lightgbm>=3.3,<4.6",  # <{N+1} upper cap, where N is the latest released minor version
     ],
     "catboost": [
-        # CatBoost wheel build is not working correctly on darwin for CatBoost 1.2, so use old version in this case.
-        # https://github.com/autogluon/autogluon/pull/3190#issuecomment-1540599280
-        # Catboost 1.2 doesn't have wheel for python 3.11
-        "catboost>=1.1,<1.2 ; sys_platform == 'darwin' and python_version < '3.11'",
-        "catboost>=1.1,<1.3; sys_platform != 'darwin'",
+        "numpy>=1.25,<2.0.0",  # TODO support numpy>=2.0.0 once issue resolved https://github.com/catboost/catboost/issues/2671
+        "catboost>=1.1,<1.3",
```
Comment (Innixma) [inline tabular/setup.py:40]: Why remove darwin logic? Does CatBoost 1.2 work on darwin now?

Comment (suzhoum) [reply tabular/setup.py:40]: Yes the support is there for Darwin, see [platform test](https://github.com/autogluon/autogluon/actions/runs/11958841532/job/33339293984#logs)
image

Comment (Innixma) [reply tabular/setup.py:40]: great thanks!

**Review by Innixma — APPROVED** [review]
LGTM! Thanks for minimizing the changes to the source code for features module!

**Pull request merged**